Stripe common settings page

Mayisha opened the pull request:
Fixes #2734 

## Changes proposed in this Pull Request:
This PR implements the same Settings screen for both non-UPE and UPE integrations. 
- For UPE enabled/disabled both scenarios, all the Stripe payment methods are listed in `<your_site>/wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=methods` page.
- For UPE enabled/disabled both scenarios, the list includes a checkbox to enable/disable the methods. The selection persists on save.
- When UPE is disabled, each method has a `Customize` button. You can add a name and a description for individual methods there except `card`.  Data persists on save.

## Testing instructions

#### Non UPE
- In `Advance settings` section of `<your_site>/wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=settings` page uncheck `New checkout experience` (disable UPE).
- In `<your_site>/wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=methods` page the methods should be listed with a checkbox. Enabling/disabling methods should work, save should work and changes should persist on page refresh.

<img width="50%" height="50%" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/33387139/0876d60f-88cd-49f1-9942-509a08500bb6" />

- Clicking on the `Customize` button should show the name and description form fields. Change some values, then click `Save changes`. This will close the section. Data gets saved when the `Save` button is clicked. 

<img width="50%" height="50%" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/33387139/0834a3f6-30b6-49ad-8de1-702cdbbdac87" />

- On `WooCommerce > Settings > Payments` page the selected payment method count should be present similar to when UPE is enabled.

![main list](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/33387139/4001333a-8267-4799-9c99-89ee32ecc89b)


#### UPE
- In `Advance settings` section of `<your_site>/wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=settings` page select `New checkout experience` (enable UPE).
- In `<your_site>/wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=methods` page the listed methods should work as before. Enabling/disabling methods should work, save should work and changes should persist on page refresh.

<img width="50%" height="50%" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/33387139/3d068bd7-57fa-4648-9e19-48ab2d29ee05" />

- Check `New checkout experience` and make sure the previously selected methods are still selected after enabling UPE.


### Note
- ToDo: add tests
- This PR does not remove the methods from `WooCommerce > Settings > Payments` page. This will be handled separately. Simply removing the classes removes them from checkout page also, so it needs further work.
- The design has the customize option for `card` too but it does not align with the current behavior and is unnecessary as the name/description has no usage for the card method).

Mayisha committed:
show legacy stripe payment method list in stripe settings page

Mayisha committed:
enable/disable legacy method from stripe settings page

Mayisha committed:
add customization component

Mayisha committed:
save individual payment gateway settings

Mayisha committed:
fix lint

Mayisha committed:
show method count

Mayisha committed:
Merge branch 'develop' into task/2734-same-settings-page

james-allan committed:
Merge branch 'develop' into task/2734-same-settings-page

james-allan commented:
I'm not sure why my local dev site appears to be broken on this branch. When I disable UPE and view the Payment methods settings screen, the top card appears to error: 

<img width="600" alt="Screenshot 2023-11-14 at 12 23 00 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/efc22bdd-7722-4536-a4d2-222b5ef0f7d9">

Enabling UPE seems to resolve the issue: 

<img width="600" alt="Screenshot 2023-11-14 at 12 25 40 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/d6fcdf47-5afa-498a-9390-178bb0d84ae0">

I don't know if this is caused by my lack of payment method options or something else. 🤷‍♂️ I also can't seem to get the payment method options to show up on my local site. 

I tested this branch on a jurassic ninja site and things worked fine there.

**UPE**

- [x] Noted the methods cannot be customised. 
- [x] Enabling/disabling methods are saved. 
- [x] Enabling/disabling methods are reflected on the checkout. 

https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/63735b96-e3a0-4127-bcc2-ce9999554a8e

**UPE Disable**

- [x] Customised payment method name and description. **
- [x] Enabling/disabling methods are saved.
- [x] Enabling/disabling methods are reflected on the checkout.  

https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/b28d90d6-2c53-45d9-9a0c-3cab5258af55

> [!NOTE]
> ** When I first customised a payment method I clicked the "save changes" button (see screenshot below) and expected that the changes would have been saved. This doesn't save, however, you have to click the "Save changes" button at the bottom of the page. This save changes button appears in the designs so this may well be the intended UX but I wanted to note it incase it is intended that the save button would save just those payment method customisations.  

<img width="400" alt="Screenshot 2023-11-14 at 2 19 23 pm" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/cbbd754d-f1ce-4066-9c95-e17a031b19b4">

Still to test: 

1. **Backwards compatibility.** For a store previously set up with Stripe, make sure checking out this branch/version doesn't change any pre-existing setup. 
2. **Code review.** I've had a glance at the changes, I haven't done a full blown review yet.

james-allan commented:
> I'm not sure why my local dev site appears to be broken on this branch. When I disable UPE and view the Payment methods settings screen, the top card appears to error:

So, just to follow up on this, I think I've been able to fix my local site by disconnecting from Stripe and re-connecting. 🤷‍♂️ That's a little awkward because it would have liked to learn what the issue was and whether there's an issue somewhere.

Mayisha committed:
add js tests

Mayisha committed:
keep enum and check upe flag

Mayisha committed:
fix php tests

Mayisha committed:
hide other stripe payment methods from settings > payments page but show on the checkout page

a-danae commented on includes/class-wc-stripe-helper.php:
```diff
@@ -337,6 +337,110 @@ public static function get_settings( $method = null, $setting = null ) {
+	public static function get_legacy_enabled_payment_methods( $field = null ) {
+		$payment_methods = self::get_legacy_payment_methods();
+
+		$enabled_payment_method_ids = [ 'card' ];
```
Looks like `card` is present in `$enabled_payment_method_ids` but not in `$enabled_payment_methods` when Cards is enabled. I'd find it confusing, but is this expected?

I get the following IDs in `$enabled_payment_method_ids`: `["card","eps","giropay","p24","sepa","sofort"]`

And the following classes in `$enabled_payment_methods`: `["WC_Gateway_Stripe_Eps","WC_Gateway_Stripe_Giropay","WC_Gateway_Stripe_P24","WC_Gateway_Stripe_Sepa","WC_Gateway_Stripe_Sofort"]` (Notice there's no class that maps to the `card` ID)

Mayisha replied:
This one is addressed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2750/commits/8f3fe940fe24d24ca42b51f931dd9b29b07bbe02

a-danae commented on includes/class-wc-gateway-stripe.php:
```diff
@@ -633,6 +643,33 @@ public function prepare_order_pay_page( $gateways ) {
+				$available_gateways[] = $legacy_enabled_gateway;
+			}
+		}
+
```
I'm seeing something unexpected on my end.

Notice some of the payment methods are duplicated in the classic checkout:
<img width="390" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/4dfec0e0-43e4-478d-b811-e315368bae58">

And on the Pay for order page:
<img width="390" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/2065dff9-241a-43df-98e4-1ec697c66e01">

When debugging this method, I'm seeing the following in `$available_gateways` and `$gateways`
<img width="390" alt="image" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/40ddc610-fdbc-415d-bc17-d43db4ec2d21">

Regarding the replication steps, I got this simply by going from `develop` to this branch. What I recall I did that _could_ be related to this behavior:

- I had UPE enabled, and then disabled it while on this branch.
- While in develop, I reordered the non-UPE payment methods at some point.

a-danae replied:
Thanks for the updates here Mayisha! These two behave as expected now:

Classic checkout page  
<img src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/ea0a30c4-59f9-44c2-aeec-290e023c95e6" width="300px" />

Pay for order page
<img src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/bb7c3a23-99ba-43ac-9594-bbd466b98e79" width="300px" />

a-danae commented on includes/class-wc-gateway-stripe.php:
```diff
@@ -633,6 +643,33 @@ public function prepare_order_pay_page( $gateways ) {
+	 */
+	public function get_payment_methods_on_checkout( $gateways ) {
+		if ( ! is_checkout() ) {
+			return $gateways;
```
Another unexpected behavior is that only cards show up in the available payment methods in these two places:

The block checkout:
<img width="390px" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/486bb74e-9d0f-4cf9-b37f-41f6e33f28da"/>

And the Add payment method page
<img width="390px" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/51856a14-e053-4d1e-b8f1-0281c17efe42"/>

When inspecting this method, it's returning at this point, and this is what `$gateways` and `$available_gateways` contain:

<img width="390px" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/d7b46387-c75f-437c-8069-f82de214f5e9"/>

a-danae replied:
> Another unexpected behavior is that only cards show up in the available payment methods in these two places:
The block checkout:

Ok, we never supported non-UPE APMs in the block checkout. As mentioned in [the critical flows](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows-using-the-classic-checkout-experience#alternative-payment-methods-apms-1). Nice catch checking this behavior persisted in `develop`, @Mayisha ! 🙂

a-danae replied:
> And the Add payment method page

Mind double-checking this one? On this page, the behavior seems to differ between develop and this branch. I enabled all of the payment gateways, and this is what I noticed:

In develop, cards and SEPA are displayed.

<img src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/fbeb51cd-9483-4cc6-af54-61ee71d61b3d" width="300px"/>

In this branch, only cards is displayed.
<img src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/41606954/ad4d857f-131f-4bad-a78d-d5311f35b702" width="300px"/>

Mayisha replied:
Thanks @a-danae. This is a very good catch. I somehow never tested with Sepa enabled and this issue went totally unnoticed on my side :flushed: 

So basically if the payment method is tokenized, it should appear on the add payment method page. I have fixed it in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2750/commits/0de57d0ca4b5ad9a58826fba308b67a31a2520bd following the logic in `woocommerce`.

a-danae replied:
Sweet, I'm no longer able to replicate the problem I described. Thanks for the iteration on this! 🙂 Marking this as resolved.

a-danae reviewed (changes requested):
Hey Mayisha! Thanks for all your work here! 🙌  

I noticed a couple of strange behaviors when continuing here with the review @james-allan did. I didn't test further to try to pinpoint what's causing this before updating the saved settings, since James didn't get this behavior. I left comments for them below.

As for this other behavior James reported...

> I'm not sure why my local dev site appears to be broken on this branch. When I disable UPE and view the Payment methods settings screen, the top card appears to error

... I didn't see it on my end just by switching branches, but I'll keep an eye on this as I keep testing.

Mayisha commented:
Thanks @a-danae for the quick look. 
I made a little [logical mistake](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/be6dacbff1a611512d7bf1848d6b21c4b8228271/includes/class-wc-gateway-stripe.php#L662-L667) in my https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2750/commits/be6dacbff1a611512d7bf1848d6b21c4b8228271 which is causing the weird data in `$available_gateways `and `$gateways`. 
This commit was pushed after @james-allan's review which is the reason he did not see these behaviours :sweat_smile: 
Fixing the issues now :eyes:

Mayisha committed:
Merge branch 'develop' into task/2734-same-settings-page

Mayisha committed:
fix logic to render legacy methods on the customer facing pages.

Mayisha committed:
check page

Mayisha committed:
fix: no enabled payment methods when switched from non-upe to upe

Mayisha committed:
enable/disable Stripe checkbox and card method following each other

Mayisha committed:
fix card getting disabled when switched from non-upe to upe

Mayisha committed:
add check

Mayisha committed:
fix condition

Mayisha committed:
add tests

Mayisha committed:
show sepa on settings page when upe is disabled
- in legacy mode the id is `stripe_sepa` and in upe the id is `sepa_debit`.
- frontend mapping and the capabilities object received from Stripe expects the id to be `sepa_debit`
- tweak the sepa id as `sepa_debit` for frontend

Mayisha committed:
fix logic of `get_available_payment_methods` method.

Mayisha committed:
fix test and undo comment out tests

a-danae commented on includes/admin/class-wc-rest-stripe-settings-controller.php:
```diff
@@ -56,65 +56,73 @@ public function register_routes() {
 						'items'             => [
 							'type' => 'string',
-							'enum' => $this->gateway->get_upe_available_payment_methods(),
+							'enum' => array_merge( $this->gateway->get_upe_available_payment_methods(), WC_Stripe_Helper::get_legacy_available_payment_method_ids() ),
```
In this other line, I see we return either the UPE or the legacy array depending on whether UPE is enabled
https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2750/files#diff-679186f0c3e522393b5394f6e65890f1b2bff5ef64efb250eee90d6348dc841eR258

So, just curious, why merge the non-UPE and UPE payment method types here? I didn't find a problem with this with a quick look. Wondering if there could be a potential inconsistency with this approach.

Mayisha replied:
I am also not sure what is the best way to handle this enum. Another option I thought was to remove this enum for the `enabled_payment_method_ids` param.

> In this other line, I see we return either the UPE or the legacy array depending on whether UPE is enabled https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2750/files#diff-679186f0c3e522393b5394f6e65890f1b2bff5ef64efb250eee90d6348dc841eR258

As this is in the function for the GET request we have a certain value for `$is_upe_enabled` flag. But here we are checking the param passed with the POST request. In this case, the saved value of `$is_upe_enabled` can differ from the param `is_upe_enabled` which is sent in the same request.

Let me know what you think about removing the enum or if you have any other suggestion. :thinking:

a-danae commented on includes/class-wc-stripe-helper.php:
```diff
@@ -337,6 +337,117 @@ public static function get_settings( $method = null, $setting = null ) {
+		$payment_method_settings = [];
+
+		foreach ( $payment_methods as $payment_method ) {
+			$settings = [
```
We're doing something similar in https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/admin/class-wc-rest-stripe-payment-gateway-controller.php#L79-L102

Do you think it'd be worth it to unify them? The keys we return are different, though, so maybe it'd require a bit of refactoring on the code that uses them. Of course, not a blocker.

a-danae replied:
I know this wasn't introduced in this PR, but wondering since we're around modifying related code... Would it be better to have a method within the $payment_method class that returns its settings and can be extended by classes that inherited it, instead of conditionally calling `get_unique_settings`?

Mayisha replied:
I have made some changes around the customization section including the `unique` settings here https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2792. I will implement this suggestion there to avoid conflicts.

a-danae commented on includes/class-wc-gateway-stripe.php:
```diff
@@ -633,6 +643,56 @@ public function prepare_order_pay_page( $gateways ) {
+			return $gateways;
+		}
+
+		$available_gateways      = [];
```
Shouldn't we add our gateways to the ones passed by this filter, instead of overriding them?

With the current approach, looks like the gateways added by other extensions are being overridden. Steps:

1. Enable UPE
2. Enable PayPal Payments (or any other extension that adds payment gateways)
3. Confirm the PayPal gateway appears on the checkout page
4. Disable UPE
5. Notice that the PayPal gateway no longer appears on the checkout page

Mayisha replied:
> Shouldn't we add our gateways to the ones passed by this filter, instead of overriding them?

You are correct. But adding our legacy gateways here is a bit tricky because we also need to maintain the ordering. The gateways passed by this filter are ordered but exclude our legacy stripe methods (as we are not loading them [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/woocommerce-gateway-stripe.php#L379) anymore to hide them from the _WooCommerce > Settings > Payments_ page. 

So we have to check the `woocommerce_gateway_order` option to push our ones to the correct position. However `woocommerce_gateway_order` option includes all the enabled/disabled methods in order. But we have to push our legacy enabled ones. This whole scenario made this a bit cofusing.

Mayisha replied:
Fixed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2750/commits/acfe1a25fdc4de50be48399c91f433ae8fa58a2c. It was happening because the payment method from the newly activated plugin was not present in the `woocommerce_gateway_order`.

a-danae replied:
Sweet! I can't replicate the problem anymore.

✅ The payment gateways from other extensions continue to appear on the checkout page when UPE is disabled

Marking this as resolved. Thanks Mayisha!

> The gateways passed by this filter are ordered but exclude our legacy stripe methods (as we are not loading them here anymore to hide them from the WooCommerce > Settings > Payments page.

Interesting. Yeah, good point. The individual gateways we add can't be ordered relative to those added by other means. Ours will always be together, either before or after other gateways. How about adding a heads-up pointing to your comment for design in the weekly update? I assume this side-effect was discussed at some point, but just in case.

a-danae commented on includes/class-wc-stripe-helper.php:
```diff
@@ -337,6 +337,117 @@ public static function get_settings( $method = null, $setting = null ) {
+	 *
+	 * @return array
+	 */
+	public static function get_legacy_enabled_payment_methods( $field = null ) {
```
Since the only field we retrieve is "id", would it be worth to reuse `get_legacy_available_payment_method_ids()` for this? The logic seems similar, and we could use an argument in that method instead of in this one to only return the enabled gateways.

a-danae reviewed (changes requested):
Thanks for the updates Mayisha! Glad to see the previous unexpected behaviors working fine 🎉 

I left some comments, but only requesting changes because of the one about the payment gateways added by other extensions being filtered out (in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2750#discussion_r1413028996). The others are just questions.

Also, how about merging this into a long-lived branch for the settings changes instead of `develop`? We'll still want to make other changes before we can release these changes, so having this in `develop` would prevent us from releasing normally.

a-danae commented on includes/class-wc-stripe-helper.php:
```diff
@@ -337,6 +337,117 @@ public static function get_settings( $method = null, $setting = null ) {
+	 * @return array
+	 */
+	public static function get_legacy_available_payment_method_ids() {
+		$payment_methods = self::get_legacy_payment_methods();
```
🤏 probably nit. Looks like we're instantiating all of the payment methods to get their IDs. I haven't found a performance problem, but maybe it'd be worth reducing the overhead for this? 

Where we instantiate them:
ehttps://github.com/woocommerce/woocommerce-gateway-stripe/pull/2750/files#diff-dbc2cfbfe4b247c27c20d86922819977adc6f7723c0b047cb9722a2f47b7ee73R368

Mayisha replied:
I made a tiny improvement in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2750/commits/7dc79b21009d3cb7f6af456d21dcc8c4febe336c.

Mayisha committed:
separate function

Mayisha committed:
reduce instantiation

Mayisha committed:
fix other plugin's payment method not appearing in checkout page

a-danae commented on includes/class-wc-stripe-helper.php:
```diff
@@ -337,6 +337,140 @@ public static function get_settings( $method = null, $setting = null ) {
+		$is_stripe_enabled = isset( $stripe_settings['enabled'] ) && 'yes' === $stripe_settings['enabled'];
+
+		$enabled_payment_methods        = self::get_legacy_enabled_payment_methods();
+		$mapped_enabled_payment_methods = array_map(
```
🤏  Wondering if it'd be worth reducing the times we iterate through the payment methods. 

For example, when we call this method, we iterate through them in 
- this method
- get_legacy_enabled_payment_methods (https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2750/files#diff-dbc2cfbfe4b247c27c20d86922819977adc6f7723c0b047cb9722a2f47b7ee73R415)
- get_legacy_payment_methods https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2750/files#diff-dbc2cfbfe4b247c27c20d86922819977adc6f7723c0b047cb9722a2f47b7ee73R378

Probably a nit, and certainly not a blocker, and no suggestions for now. Leaving my thoughts here in case it's worth actioning later.

a-danae reviewed (approved):
🎉 Thanks for all the work here! Things look good to me.

There are a couple of open conversations in the comments (https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2750#discussion_r1419289056 and https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2750#discussion_r1409573646), but we can continue without blocking this PR. Glad to see these changes merged 🙂

Mayisha commented:
Thanks @a-danae. :smiley_cat: 

I will wrap up and address the open conversations in one of the upcoming PRs. I have created a feature branch `task/updated-settings-page` and I am going to merge this PR to that branch.
